code/hsec-sync/src/Security/Advisories/Sync/Snapshot.hs

```diff
@@ -0,0 +1,181 @@
+import Control.Monad.Trans.Except (runExceptT, throwE)
+import Data.Aeson (FromJSON, eitherDecodeFileStrict)
+import qualified Data.ByteString.Lazy as BL
+import Data.Either.Combinators (whenLeft)
+import Data.Functor (($>))
+import Data.Time (UTCTime)
+import GHC.Generics (Generic)
+import Network.HTTP.Client (HttpException (..), HttpExceptionContent (..))
+import Network.Wreq
+import qualified System.Directory as D
+import System.FilePath ((</>))
+import System.IO.Temp (withSystemTempDirectory)
+import Security.Advisories.Sync.Url
+
+data SnapshotError
+  = SnapshotDirectoryMissingE
+  | SnapshotIncoherent String
+  | SnapshotProcessError SnapshotProcessError
+
+data SnapshotProcessError
+  = FetchSnapshotArchive String
+  | DirectorySetupSnapshotArchive IOException
+  | ExtractSnapshotArchive IOException
+
+explainSnapshotError :: SnapshotError -> String
+explainSnapshotError =
+  \case
+    SnapshotDirectoryMissingE -> "Snapshot directory is missing"
+    SnapshotIncoherent e -> "Snapshot directory is incoherent: " <> e
+    SnapshotProcessError e ->
+      unlines
+        [ "An exception occurred during snapshot processing:"
+        , case e of
+            FetchSnapshotArchive x -> "Fetching failed with: " <> show x
+            DirectorySetupSnapshotArchive x -> "Directory setup got an exception: " <> show x
+            ExtractSnapshotArchive x -> "Extraction got an exception: " <> show x
+        ]
+
+data Snapshot = Snapshot
+  { snapshotRoot :: FilePath,
+    repositoryUrl :: String,
+    repositoryBranch :: String
+  }
+
+snapshotArchiveUrl :: Snapshot -> String
+snapshotArchiveUrl s =
+  ensureFile (mkUrl [repositoryUrl s, "archive/refs/heads", repositoryBranch s]) <> ".tar.gz"
+
+data SnapshotRepositoryStatus
+  = SnapshotDirectoryMissing
+  | SnapshotDirectoryInitialized
+  | SnapshotDirectoryIncoherent
+
+snapshotRepositoryStatus :: Snapshot -> IO SnapshotRepositoryStatus
+snapshotRepositoryStatus s = do
+  dirExists <- D.doesDirectoryExist $ snapshotRoot s
+  if dirExists
+    then do
+      dirAdvisoriesExists <- D.doesDirectoryExist $ snapshotRoot s </> "advisories"
+      fileMetadataExists <- D.doesFileExist $ snapshotRoot s </> "snapshot.json"
+      return $
+        if dirAdvisoriesExists && fileMetadataExists
+          then SnapshotDirectoryInitialized
+          else SnapshotDirectoryIncoherent
+    else return SnapshotDirectoryMissing
+
+data SnapshotRepositoryEnsuredStatus
+  = SnapshotRepositoryCreated
+  | SnapshotRepositoryExisting
+
+ensureSnapshot ::
+  Snapshot ->
+  SnapshotRepositoryStatus ->
+  IO (Either SnapshotError SnapshotRepositoryEnsuredStatus)
+ensureSnapshot s =
+  \case
+    SnapshotDirectoryMissing ->
+      ($> SnapshotRepositoryCreated)
+        <$> overwriteSnapshot s
+    SnapshotDirectoryIncoherent ->
+      ($> SnapshotRepositoryCreated)
+        <$> overwriteSnapshot s
+    SnapshotDirectoryInitialized ->
+      return $ Right SnapshotRepositoryExisting
```

Comment:
if this was `ExceptT IO`, it would be less awkward.

Reply:
It's a plain `case` with two lines each, not a long procedure with nested conditionals, using `ExceptT` here would be overkill.

Reply:
you get an `ExceptT` from `overwriteSnapShot`, if you don't run it and you save a hole lot of `fmap`ing, I don't know how it's overkill, you just need to move the `runExceptT` to this function.